addons/mail/controllers/discuss_controller.py

```diff
@@ -0,0 +1,284 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+import base64
+import json
+import logging
+import unicodedata
+
+_logger = logging.getLogger(__name__)
+
+from odoo import http
+from odoo.addons.web.controllers.main import Binary, clean, serialize_exception
+from odoo.exceptions import UserError
+from odoo.http import request
+from odoo.osv import expression
+from odoo.tools import consteq
+from odoo.tools.translate import _
+from werkzeug.exceptions import NotFound
+
+class DiscussController(http.Controller):
+
+    @http.route('/discuss/channel/<int:channel_id>', type='http', methods=['GET'], auth='public')
+    def discuss_channel(self, channel_id, **kwargs):
+        # Ensure the current user or guest is member of the channel.
+        channel_partner = self._get_channel_partner(channel_id)
+        if not channel_partner:
+            raise NotFound()
+
+        return request.render('mail.channel_template', {
+            'channel': channel_partner.channel_id,
+            # needed to initialize the web client
+            'session_info': channel_partner.env['ir.http'].session_info(),
+        })
+    
+    @http.route('/chat/<int:channel_id>/<string:invitation_token>', methods=['GET'], auth='public')
+    def join_channel(self, channel_id, invitation_token, **kwargs):
+        channel = self._get_channel(channel_id)
+        if not channel or not consteq(channel.uuid, invitation_token):
+            raise NotFound()
+
+        response = request.redirect(f'/discuss/channel/{channel.id}')
+
+        # current user is already member of the channel: redirect it to the channel
+        if self._get_channel_partner(channel_id):
+            return response
+
+        if request.session.uid:
+            channel.add_members([request.env.user.partner_id])
+        else:
+            guest = request.env['mail.guest']._get_or_create_guest_from_request(request, response)
```

Comment:
maybe it's just me but I find it weird that this function whose name starts with _get has an important side-effect such as setting cookies on the response, or that it takes a response as an argument in the first place when it's named "from_request".

Considering this is the only usage of that method, I would do the cookie setting in this controller instead.